### PR TITLE
Update Dockerfiles to be multi-staged

### PR DESCRIPTION
**Pull request by MichaelSimons**
Related to #16 

This reduces the image sizes substantially.
Linux => 1.81GB to 398 MB
Windows => 2.83GB to 1.14 GB

### CODE REVIEW

File: src/Microsoft.DotNet.ImageBuilder/Dockerfile [hunk 1]
```diff
@@ -2,12 +2,29 @@
...
 #     docker run --rm -v /var/run/docker.sock:/var/run/docker.sock -v <local path to build>:/repo -w /repo image-builder <image-build args>
 
-FROM microsoft/dotnet:2.0-sdk
+FROM microsoft/dotnet:2.0-sdk AS build-env
 
-# Install Docker
+# build Microsoft.DotNet.ImageBuilder
+WORKDIR image-builder
+
+# copy csproj and restore as distinct layers
```
Comment (dagood) [inline src/Microsoft.DotNet.ImageBuilder/Dockerfile]: When I read this at first I thought it meant copying the csproj being distinct from the `dotnet restore`, but it really means this group of three commands being distinct from copying the source files, right? (For dev rebuild efficiency.) Seems the comment could be clarified.

File: src/Microsoft.DotNet.ImageBuilder/Dockerfile [hunk 2]
```diff
@@ -2,12 +2,29 @@
...
+
+
+# build runtime image
+FROM microsoft/dotnet:2.0-runtime
+
+# install Docker
 RUN apt-get update \
     && apt-get install -y --no-install-recommends \
         apt-transport-https \
+        gnupg \
```
Comment (dagood) [inline src/Microsoft.DotNet.ImageBuilder/Dockerfile]: Just curious what this is for.

Comment (MichaelSimons) [inline src/Microsoft.DotNet.ImageBuilder/Dockerfile]: This is a docker-ce prereq.  Before this change it was coming from the base image chain.  specifically it is part of the buildpack-deps image which is the base image for the .NET Core SDK image

File: src/Microsoft.DotNet.ImageBuilder/Dockerfile (hunk 1 above)
Comment (MichaelSimons) [inline src/Microsoft.DotNet.ImageBuilder/Dockerfile]: Is this better?  `restore packages before coping entire source - provides optimizations when rebuilding`

Comment (dagood) [inline src/Microsoft.DotNet.ImageBuilder/Dockerfile]: Perfect, with `coping` -> `copying`. :smile:

**Pull request merged**